src/basic/path-util.c

```diff
@@ -1172,3 +1173,112 @@ bool credential_name_valid(const char *s) {
          * them around) and as fdnames (which is how we might want to pass them around eventually) */
         return filename_is_valid(s) && fdname_is_valid(s);
 }
+
+/* Unfortunately *ftw does not take a void *userdata pointer, so
+ * we have to use global variables to pass data... */
+struct nftw_cb_data {
+        const char *root_path;
+        char ***mounts;
+        char ***overlays;
+        size_t extra_last;
+};
+static thread_local struct nftw_cb_data *cb_data;
+
+static int nftw_cb(
+                const char *fpath,
+                const struct stat *sb,
+                int tflag,
+                struct FTW *ftwbuf) {
+
+        _cleanup_free_ char *parent_relative = NULL;
+        const char *parent, *check;
+        struct stat st;
+        int r;
+
+        if (tflag != FTW_F && tflag != FTW_D)
+                return 0;
+
+        if (isempty(&fpath[cb_data->extra_last]))
+                return 0;
+
+        /* Ignore /usr and /opt themselves, we only care about children */
+        if (STR_IN_SET(&fpath[cb_data->extra_last], "/usr", "/opt"))
+                return 0;
+        /* Ignore anything not below /usr/ and /opt/ */
+        if (path_startswith_strv(&fpath[cb_data->extra_last], STRV_MAKE("/usr/", "/opt/")) == NULL)
+                return 0;
+        /* Already added its parent */
+        if (path_startswith_strv(&fpath[cb_data->extra_last], *cb_data->mounts) != NULL)
+                return 0;
+        if (path_startswith_strv(&fpath[cb_data->extra_last], *cb_data->overlays) != NULL)
+                return 0;
+
+        check = prefix_roota(cb_data->root_path, &fpath[cb_data->extra_last]);
+        r = access(check, F_OK);
+        if (r == 0) {
+                /* file/dir exists in the root */
```

Comment:
you need to take the mounts from b_data->mounts into account here... as any earlier layer might have the dir, not just the base...

i.e. ->mounts should probably be a hashmap keyed by the path, returning the dir it shall be bind mounted from.

Reply:
uhm the way it's intended to use, as discussed yesterday, is to compare the fully applied view with the extension - ie, (((root + a) + b) + c), so the base will have everything already when this comparison is done

Or am I misunderstanding the point?